File: An2/MPP/mpp-proiect-java-DeepSmeag/persistenceModule/src/main/java/contests/persistence/ParticipantRepo.java
package contests.persistence;

import contests.model.Participant;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import contests.persistence.interfaces.ParticipantRepoInterface;
import contests.persistence.utils.DBConnection;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class ParticipantRepo implements ParticipantRepoInterface {

    private static final Logger logger = LogManager.getLogger();

    @Override
    public boolean saveEntity(Participant entity) {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("INSERT INTO \"Participanti\" (\"nameP\", \"age\") VALUES(?, ?)")) {
            ps.setString(1, entity.getNume());
            ps.setInt(2, entity.getVarsta());
            ps.executeUpdate();
//                Log.d("tag", "restul");
            return true;
        } catch (SQLException e) {
//            throw new RuntimeException(e);
//            Log.E("ErRROR","ERR");
            return false;
        }
    }

    @Override
    public boolean deleteEntity(Integer id) {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("DELETE FROM \"Participanti\" WHERE \"id\" = ?")) {
            ps.setInt(0, id);
            ps.executeUpdate();
//                Log.d("tag", "restul");
            return true;
        } catch (SQLException e) {
//            throw new RuntimeException(e);
//            Log.E("ErRROR","ERR");
            return false;
        }
    }

    @Override
    public boolean updateEntity(Participant entity) {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("UPDATE \"Participanti\" SET \"nameP\" = ?,  \"age\" = ? WHERE \"id\" = ?")) {
            ps.setString(0, entity.getNume());
            ps.setInt(1, entity.getVarsta());
            ps.setInt(2, entity.getId());
            ps.executeUpdate();
//                Log.d("tag", "restul");
            return true;
        } catch (SQLException e) {
//            throw new RuntimeException(e);
//            Log.E("ErRROR","ERR");
            return false;
        }
    }

    @Override
    public List<Participant> getAll() {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT * FROM \"Participanti\"")) {
            ResultSet resultSet = ps.executeQuery();
            ArrayList<Participant> participants = new ArrayList<>();
            while (resultSet.next()) {
                Integer id = resultSet.getInt("id");
                String name = resultSet.getString("nameP");
                int age = resultSet.getInt("age");
                Participant participant = new Participant(id, name, age);
//                Log.d("tag", "restul");
                participants.add(participant);
            }
            return participants;
        } catch (SQLException e) {
//            throw new RuntimeException(e);
//            Log.E("ErRROR","ERR");
            return null;
        }
    }

    @Override
    public Participant getOne(Integer id) {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT * FROM \"Participanti\" WHERE id=?")) {
            ps.setInt(0, id);
            ResultSet resultSet = ps.executeQuery();
            resultSet.next();
            String name = resultSet.getString("nameP");
            int age = resultSet.getInt("age");
            Participant participant = new Participant(id, name, age);
//                Log.d("tag", "restul");


            return participant;
        } catch (SQLException e) {
//            throw new RuntimeException(e);
//            Log.E("ErRROR","ERR");
            return null;
        }
    }

    @Override
    public Participant getParticipantByName(String name) {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT * FROM \"Participanti\" WHERE \"nameP\"=?")) {
            ps.setString(1, name);
            ResultSet resultSet = ps.executeQuery();
            resultSet.next();
            Integer id = resultSet.getInt("id");
            int age = resultSet.getInt("age");
            Participant participant = new Participant(id, name, age);
//                Log.d("tag", "restul");


            return participant;
        } catch (SQLException e) {
//            throw new RuntimeException(e);
//            Log.E("ErRROR","ERR");
            return null;
        }
    }

    @Override
    public List<Participant> getParticipantsByAgeCategory(int ageMin, int ageMax) {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT * FROM \"Participanti\" WHERE \"age\"<=? AND \"age\">=?")) {
            ps.setInt(0, ageMin);
            ps.setInt(1, ageMax);
            ResultSet resultSet = ps.executeQuery();
            ArrayList<Participant> participants = new ArrayList<>();
            while (resultSet.next()) {
                Integer id = resultSet.getInt("id");
                String name = resultSet.getString("nameP");
                int age = resultSet.getInt("age");
                Participant participant = new Participant(id, name, age);
//                Log.d("tag", "restul");
                participants.add(participant);
            }
            return participants;
        } catch (SQLException e) {
//            throw new RuntimeException(e);
//            Log.E("ErRROR","ERR");
            return null;
        }
    }

    public List<Participant> getParticipantsByProba(int idProba) {
        try (Connection connection = DBConnection.getNewConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT \"P\".\"id\", \"P\".\"nameP\", \"P\".age FROM \"Participanti\" \"P\" INNER JOIN \"Inscrieri\" ON \"P\".\"id\"=\"Inscrieri\".\"idParticipant\" WHERE \"Inscrieri\".\"idProba\"=?")) {
            ps.setInt(1, idProba);
            ResultSet resultSet = ps.executeQuery();
            ArrayList<Participant> participants = new ArrayList<>();
            while (resultSet.next()) {
                Integer id = resultSet.getInt("id");
                String name = resultSet.getString("nameP");
                int age = resultSet.getInt("age");
                Participant participant = new Participant(id, name, age);
                participants.add(participant);
            }
            return participants;
        } catch (SQLException e) {
//            throw new RuntimeException(e);
            logger.error("Error", e);
        }
        return null;
    }
}

File: An2/MPP/mpp-proiect-csharp-DeepSmeag/persistence/ParticipantRepo.cs
using mpp_proiect_csharp_DeepSmeag.models;
using mpp_proiect_csharp_DeepSmeag.persistence.interfaces;
using mpp_proiect_csharp_DeepSmeag.persistence.utils;
using Npgsql;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace mpp_proiect_csharp_DeepSmeag.persistence
{
    public class ParticipantRepo : IParticipantRepoInterface
    {
        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
        public bool deleteEntity(int entityId)
        {
            try
            {
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "DELETE FROM \"Participanti\" WHERE \"id\"=@id";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                sqlCommand.Parameters.AddWithValue("id", entityId);
                sqlCommand.Prepare();
                sqlCommand.ExecuteNonQuery();
                con.Close();
                return true;
            }
            catch (Exception e)
            {
                return false;
            }
        }

        public List<Participant>? getAll()
        {
            try
            {
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "SELECT * FROM \"Participanti\"";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                using var reader = sqlCommand.ExecuteReader();
                var participanti = new List<Participant>();
                while (reader.Read())
                {
                    var participant = new Participant(reader.GetInt32(0), reader.GetString(1), reader.GetInt32(2));
                    participanti.Add(participant);
                }
                con.Close();
                return participanti;
            }
            catch (Exception e)
            {
                return null;
            }
        }

        public Participant? getOne(int entityId)
        {
            try
            {
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "SELECT * FROM \"Participanti\" WHERE \"id\"=@id";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                sqlCommand.Parameters.AddWithValue("id", entityId);
                sqlCommand.Prepare();
                using var reader = sqlCommand.ExecuteReader();
                if (reader.Read())
                {
                    var participant = new Participant(reader.GetInt32(0), reader.GetString(1), reader.GetInt32(2));
                    con.Close();
                    return participant;
                }
                con.Close();
                return null;
            }
            catch (Exception e)
            {
                return null;
            }
        }

        public Participant? getParticipantByName(string name)
        {
            try
            {
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "SELECT * FROM \"Participanti\" WHERE \"nameP\"=@nume";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                sqlCommand.Parameters.AddWithValue("nume", name);
                sqlCommand.Prepare();
                using var reader = sqlCommand.ExecuteReader();
                if (reader.Read())
                {
                    var participant = new Participant(reader.GetInt32(0), reader.GetString(1), reader.GetInt32(2));
                    con.Close();
                    return participant;
                }
                con.Close();
                return null;
            }
            catch (Exception e)
            {
                return null;
            }
        }

        public List<Participant>? getParticipantsByAgeCategory(int ageMin, int ageMax)
        {
            try
            {
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "SELECT * FROM \"Participanti\" WHERE \"age\" BETWEEN @ageMin AND @ageMax";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                sqlCommand.Parameters.AddWithValue("ageMin", ageMin);
                sqlCommand.Parameters.AddWithValue("ageMax", ageMax);
                sqlCommand.Prepare();
                using var reader = sqlCommand.ExecuteReader();
                var participanti = new List<Participant>();
                while (reader.Read())
                {
                    var participant = new Participant(reader.GetInt32(0), reader.GetString(1), reader.GetInt32(2));
                    participanti.Add(participant);
                }
                con.Close();
                return participanti;
            }
            catch (Exception e)
            {
                return null;
            }
        }

        public bool saveEntity(Participant entityType)
        {
            try
            {
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "INSERT INTO \"Participanti\" (\"nameP\", \"age\") VALUES (@nameP, @age)";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                sqlCommand.Parameters.AddWithValue("nameP", entityType.Nume);
                sqlCommand.Parameters.AddWithValue("age", entityType.Varsta);
                sqlCommand.Prepare();
                sqlCommand.ExecuteNonQuery();
                con.Close();
                return true;
            }
            catch (Exception e)
            {
                return false;
            }
        }

        public bool updateEntity(Participant entityType)
        {
            try
            {
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "UPDATE \"Participanti\" SET \"nameP\"=@nameP, \"age\"=@age WHERE \"id\"=@id";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                sqlCommand.Parameters.AddWithValue("nameP", entityType.Nume);
                sqlCommand.Parameters.AddWithValue("age", entityType.Varsta);
                sqlCommand.Parameters.AddWithValue("id", entityType.Id);
                sqlCommand.Prepare();
                sqlCommand.ExecuteNonQuery();
                con.Close();
                return true;
            }
            catch (Exception e)
            {
                return false;
            }
        }
        public List<Participant> getParticipantsByProba(int idProba)
        {
            try
            {
                using var con = DBUtils.GetNewConnection();
                con.Open();
                var sqlString = "SELECT \"P\".\"id\", \"P\".\"nameP\", \"P\".age FROM \"Participanti\" \"P\" INNER JOIN \"Inscrieri\" ON \"P\".\"id\"=\"Inscrieri\".\"idParticipant\" WHERE \"Inscrieri\".\"idProba\"=@idProba";
                using var sqlCommand = new NpgsqlCommand(sqlString, con);
                sqlCommand.Parameters.AddWithValue("idProba", idProba);
                sqlCommand.Prepare();
                using var reader = sqlCommand.ExecuteReader();
                var participanti = new List<Participant>();
                while (reader.Read())
                {
                    var participant = new Participant(reader.GetInt32(0), reader.GetString(1), reader.GetInt32(2));
                    participanti.Add(participant);
                }
                con.Close();
                return participanti;
            }
            catch (Exception e)
            {
                return null;
            }
        }

      
    }
}
